### PR TITLE
Testsuite: Add test for salt minion ssh push via tunnel

### PR DESCRIPTION
**Pull request by calancha**
## What does this PR change?
Add a test for SSH push via tunnel (Salt minion)

After botstrap it, install/remove a package on it and
run a remote command.
- features/core/srv_create_activationkey.feature:
Create a test key for SSH push via tunnel.

- testsuite/features/secondary/min_ssh_tunnel.feature:
New test.
- testsuite/run_sets/secondary.yml:
Add it to the secondary stage.

- testsuite/features/support/twopence_init.rb:
Add label 'sle_ssh_tunnel_minion'.
- testsuite/documentation/cucumber-steps.md:
Document the new label.

- testsuite/features/secondary/trad_ssh_push.feature:
Do not create new keys for ssh/ssh-tunnel, use the existent ones.

## Links
https://github.com/SUSE/spacewalk/issues/11427
### Ports
- Manager-4.0: https://github.com/SUSE/spacewalk/pull/12839
- Manager-4.1: https://github.com/SUSE/spacewalk/pull/12838

## Changelogs

If you don't need a changelog check, please mark this checkbox:

- [x] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

### CODE REVIEW

File: testsuite/features/secondary/min_ssh_push.feature [hunk 1]
```diff
@@ -0,0 +1,48 @@
+# Copyright (c) 2020 SUSE LLC
+# Licensed under the terms of the MIT license.
+
+Feature: Register a salt system to be managed via SSH push
```
Comment (srbarrios) [inline testsuite/features/secondary/min_ssh_push.feature:4]: I would prefer to use the same naming everywhere, not sure which is the better one, I would say "via SSH Tunnel" ?

File: testsuite/features/secondary/min_ssh_push.feature [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  Scenario: Delete the salt minion for ssh-reverse bootstrap
+    Given I am on the Systems overview page of this "sle_minion"
+    When I follow "Delete System"
+    Then I should see a "Confirm System Profile Deletion" text
+    When I click on "Delete Profile"
+    And I wait until I see "has been deleted" text
+    And I cleanup minion "sle_minion"
+    Then "sle_minion" should not be registered
+
+  Scenario: Register this minion for SSH push via tunnel
```
Comment (srbarrios) [inline testsuite/features/secondary/min_ssh_push.feature:15]: for push via ssh tunnel maybe?

File: testsuite/features/secondary/min_ssh_push.feature [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+  Scenario: Cleanup: delete the SSH push tunnel minion
+    Given I am on the Systems overview page of this "sle_minion"
+    When I follow "Delete System"
+    Then I should see a "Confirm System Profile Deletion" text
+    When I click on "Delete Profile"
+    And I wait until I see "has been deleted" text
+    Then "sle_minion" should not be registered
+
+  Scenario: Cleanup: register a salt minion after SSH push tests
```
Comment (srbarrios) [inline testsuite/features/secondary/min_ssh_push.feature:44]: :D Definitely we need to find a unique name to describe that kind of test

File: testsuite/features/secondary/min_ssh_push.feature [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    And I select "1-SUSE-SSH-TUNNEL-DEV-x86_64" from "activationKeys"
+    And I select the hostname of "proxy" from "proxies"
+    And I check "manageWithSSH"
+    And I click on "Bootstrap"
+    And I wait until I see "Successfully bootstrapped host!" text
+    And I am on the System Overview page
+    And I wait until I see the name of "sle_minion", refreshing the page
+    And I wait until onboarding is completed for "sle_minion"
+
+  Scenario: Check this minion's contact method
```
Comment (srbarrios) [inline testsuite/features/secondary/min_ssh_push.feature:32]: I would use a more unique name for that scenario, as one of our agreements is not repeat scenario names, and in this case can be to generic.

File: testsuite/features/secondary/min_ssh_push.feature [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    When I click on "Delete Profile"
+    And I wait until I see "has been deleted" text
+    And I cleanup minion "sle_minion"
+    Then "sle_minion" should not be registered
+
+  Scenario: Register this minion for SSH push via tunnel
+    Given I am authorized
+    And I go to the bootstrapping page
+    Then I should see a "Bootstrap Minions" text
+    And I enter the hostname of "sle_minion" as "hostname"
```
Comment (srbarrios) [inline testsuite/features/secondary/min_ssh_push.feature:19]: When

File: testsuite/features/secondary/min_ssh_push.feature [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    And I enter "22" as "port"
+    And I enter "root" as "user"
+    And I enter "linux" as "password"
+    And I select "1-SUSE-SSH-TUNNEL-DEV-x86_64" from "activationKeys"
+    And I select the hostname of "proxy" from "proxies"
+    And I check "manageWithSSH"
+    And I click on "Bootstrap"
+    And I wait until I see "Successfully bootstrapped host!" text
```
Comment (srbarrios) [inline testsuite/features/secondary/min_ssh_push.feature:27]: Then

File: testsuite/features/secondary/min_ssh_push.feature [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    And I am on the System Overview page
+    And I wait until I see the name of "sle_minion", refreshing the page
```
Comment (srbarrios) [inline testsuite/features/secondary/min_ssh_push.feature:29]: Do we really need those two steps? Maybe is enough with the last step `I wait until onboarding is completed for "sle_minion"`

**Comment by srbarrios** [comment]
Take a look to this comment please: https://github.com/SUSE/spacewalk/issues/11370#issuecomment-633624908

It would be nice to test remote command and installation of a package using salt minion via ssh tunnel

File: testsuite/features/core/srv_create_activationkey.feature [hunk 8]
```diff
@@ -119,3 +119,15 @@ Feature: Be able to create and manipulate activation keys
     And I select "Push via SSH" from "contact-method"
     And I click on "Create Activation Key"
     Then I should see a "Activation key SUSE SSH Test Key x86_64 has been created" text
+
```
Comment (Bischoff) [inline testsuite/features/core/srv_create_activationkey.feature:122]: Please ` Copyright (c) 2010-2020 SUSE LLC` on top of file.

File: testsuite/features/secondary/min_ssh_push.feature [hunk 9]
```diff
@@ -0,0 +1,48 @@
+# Copyright (c) 2020 SUSE LLC
+# Licensed under the terms of the MIT license.
+
+Feature: Register a salt system to be managed via SSH push
+
+  Scenario: Delete the salt minion for ssh-reverse bootstrap
+    Given I am on the Systems overview page of this "sle_minion"
```
Comment (Bischoff) [inline testsuite/features/secondary/min_ssh_push.feature:7]: We had a convention (introduced by Dario Maiocchi) to always use a minion name that reflects client's role. In this case "sle_minion" would also be called "sle_reverse_ssh_minion" or something alike.

I'm not a big fan of this convention and we could also drop it alltogether, but then we need to also remove it at the other places.

If not, we need to conform to it.

File: testsuite/features/secondary/min_ssh_push.feature (hunk 2 above)
Comment (Bischoff) [reply testsuite/features/secondary/min_ssh_push.feature:15]: via SSH tunnel then (capital, it's the protocol)

File: testsuite/features/secondary/min_ssh_push.feature (hunk 6 above)
Comment (Bischoff) [reply testsuite/features/secondary/min_ssh_push.feature:27]: Not necessarily. This is an action (to wait). But ok.

File: testsuite/features/secondary/min_ssh_push.feature [hunk 10]
```diff
@@ -0,0 +1,48 @@
+# Copyright (c) 2020 SUSE LLC
+# Licensed under the terms of the MIT license.
+
+Feature: Register a salt system to be managed via SSH push
+
+  Scenario: Delete the salt minion for ssh-reverse bootstrap
```
Comment (Bischoff) [inline testsuite/features/secondary/min_ssh_push.feature:6]: what is "ssh-reverse" ? the name of a program?

If it's just "SSH reverse", then no dash, and use capital letters for protocol name.

File: testsuite/features/secondary/trad_ssh_push.feature [hunk 11]
```diff
@@ -3,36 +3,16 @@
 
 Feature: Register a traditional system to be managed via SSH push
 
-  Scenario: Delete the traditional client for ssh-reverse bootrap
+  Scenario: Delete the traditional client for ssh-reverse bootstrap
```
Comment (Bischoff) [inline testsuite/features/secondary/trad_ssh_push.feature:6]: nice catch.

(SSH reverse ?)

File: testsuite/features/secondary/trad_ssh_push.feature [hunk 12]
```diff
@@ -3,36 +3,16 @@
...
-  Scenario: Create an activation key for SSH push via tunnel
-    Given I am on the Systems page
-    When I follow the left menu "Systems > Activation Keys"
-    And I follow "Create Key"
-    And I enter "SSH push via tunnel key" as "description"
-    And I enter "ssh-push-tunnel" as "key"
-    And I select "Push via SSH tunnel" from "contact-method"
-    And I click on "Create Activation Key"
-    Then I should see a "Activation key SSH push via tunnel key has been created." text
-
```
Comment (Bischoff) [inline testsuite/features/secondary/trad_ssh_push.feature:33]: nice catch.

**Review by Bischoff — APPROVED** [review]
Nice work. See nitpicks.

File: testsuite/features/secondary/min_ssh_push.feature (hunk 9 above)
Comment (srbarrios) [reply testsuite/features/secondary/min_ssh_push.feature:7]: let's keep the convention, please! But what about some shorter name? "ssh_push_minion"?

**Review by srbarrios — CHANGES_REQUESTED** [review]
Take a look to this comment please: SUSE/spacewalk#11370 (comment)

It would be nice to test remote command and installation of a package using salt minion via ssh tunnel. As the installation of a package through ssh-tunnel is the key-piece of this feature.

File: testsuite/features/secondary/min_ssh_push.feature (hunk 9 above)
Comment (calancha) [reply testsuite/features/secondary/min_ssh_push.feature:7]: Ok, I will go with `ssh_push_minion`

File: testsuite/features/secondary/min_ssh_push.feature (hunk 4 above)
Comment (calancha) [reply testsuite/features/secondary/min_ssh_push.feature:32]: `Scenario: The contact method is SSH tunnel on this minion`?

File: testsuite/features/secondary/min_ssh_push.feature (hunk 6 above)
Comment (calancha) [reply testsuite/features/secondary/min_ssh_push.feature:27]: I'll follow Eric here (i.e. keep `And`).

File: testsuite/features/secondary/min_ssh_push.feature (hunk 7 above)
Comment (calancha) [reply testsuite/features/secondary/min_ssh_push.feature:29]: You are right.  Fixed.

File: testsuite/features/core/srv_create_activationkey.feature (hunk 8 above)
Comment (calancha) [reply testsuite/features/core/srv_create_activationkey.feature:122]: Copyright should start with the year of publication of the work (this file in this case).  The copyright cannot go backward in time before the existence of the work.

**Comment by calancha** [comment]
@srbarrios Added install/remove a package tests and test for remote command.

File: testsuite/features/secondary/min_ssh_push.feature (hunk 9 above)
Comment (srbarrios) [reply testsuite/features/secondary/min_ssh_push.feature]: After review doc, maybe "ssh push" is not the best way to identify this client.
See that https://documentation.suse.com/external-tree/en-us/suma/4.1/suse-manager/architecture/salt-contact-method-overview.html

image

image

So... maybe is better to change in the feature file name and in the host name by something else like "ssh_tunnel_minion", sorry for all this confusion...

**Review by lkotek — APPROVED** [review]
I will be great to have it, thanks :+1:

**Review by srbarrios — APPROVED** [review]
After we change the naming. Looks all good to me.

File: testsuite/features/secondary/min_ssh_tunnel.feature [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+
+  Scenario: Cleanup: delete the SSH push tunnel minion
+    Given I am on the Systems overview page of this "sle_ssh_tunnel_minion"
+    When I follow "Delete System"
+    Then I should see a "Confirm System Profile Deletion" text
+    When I click on "Delete Profile"
+    And I wait until I see "has been deleted" text
+    Then "sle_ssh_tunnel_minion" should not be registered
+
+  Scenario: Cleanup: register a salt minion after SSH push tests
```
Comment (srbarrios) [inline testsuite/features/secondary/min_ssh_tunnel.feature:82]: No need to fix.. but just so you are aware, maybe here we should also say SSH tunnel minion

File: testsuite/features/secondary/min_ssh_tunnel.feature [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+    And I enter "milkyway-dummy" as the filtered package name
+    And I click on the filter button
+    And I check "milkyway-dummy" in the list
+    And I click on "Remove Packages"
+    And I click on "Confirm"
+    Then I should see a "1 package removal has been scheduled" text
+    When I force picking pending events on "sle_ssh_tunnel_minion" if necessary
+    Then I wait until event "Package Removal scheduled by admin" is completed
+
+  Scenario: Run a remote command on this SSH reverse minion
```
Comment (srbarrios) [inline testsuite/features/secondary/min_ssh_tunnel.feature:60]: No need to fix.. but just so you are aware, maybe here we should also say SSH tunnel minion

**Pull request merged**